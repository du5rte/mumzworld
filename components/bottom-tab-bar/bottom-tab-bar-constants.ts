export const TAB_PADDING = 8;
export const TAB_SIZE = 48;
export const TAB_ICON_SIZE = 28;
export const TAB_BAR_HEIGHT = TAB_SIZE + TAB_PADDING * 2; // 64
export const TAB_BAR_MARGIN = 12;
export const TAB_BAR_MAX_WIDTH = 408;
