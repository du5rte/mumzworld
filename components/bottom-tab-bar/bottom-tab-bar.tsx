import React, { useMemo } from 'react';
import { I18nManager, LayoutChangeEvent, TouchableOpacity, View } from 'react-native';
import { useSafeAreaInsets } from 'react-native-safe-area-context';
import { BottomTabBarProps } from '@react-navigation/bottom-tabs/src/types';
import Feather from '@expo/vector-icons/Feather';
import Animated, { useAnimatedStyle, useSharedValue } from 'react-native-reanimated';
import useTheme from '@/hooks/useTheme';
import useBottomTabBarShown from '@/hooks/useBottomTabBarShownAtom';
import { withEaseOutQuad, withEaseOutSin } from '@/styles/timings';
import {
  TAB_SIZE,
  TAB_BAR_HEIGHT,
  TAB_BAR_MARGIN,
  TAB_ICON_SIZE,
} from './bottom-tab-bar-constants';
import Circle from '../circle';
import Box from '../box';

const AnimatedBox = Animated.createAnimatedComponent(Box);

export type FeatherIconName = keyof typeof Feather.glyphMap;

export interface Tab {
  id: string;
  label: string;
  icon: FeatherIconName;
  onPress?: () => void;
  onLongPress?: () => void;
}

export interface TabItemProps extends Tab {
  id: string;
  label: string;
  icon: FeatherIconName;
  focused: boolean;
  onPress: () => void;
  onLongPress: () => void;
}

export function TabItem(props: TabItemProps) {
  const { label, icon, focused, onPress, onLongPress } = props;

  const theme = useTheme();

  return (
    <TouchableOpacity
      accessibilityRole="button"
      accessibilityState={focused ? { selected: true } : {}}
      accessibilityLabel={label}
      testID={`tab-item-${props.id}${focused ? '-focused' : ''}`}
      onPress={onPress}
      onLongPress={onLongPress}>
      <Circle size={TAB_SIZE}>
        <Feather
          name={icon}
          size={TAB_ICON_SIZE}
          color={focused ? theme.colors.primary : theme.colors.inactive}
        />
      </Circle>
    </TouchableOpacity>
  );
}

export interface TabBarProps {
  tabs: Tab[];
  currentTabIndex: number;
  onTabPress?: (index: number, tabs: Tab) => void;
  onTabLongPress?: (index: number, tabs: Tab) => void;
}

export function TabBar(props: TabBarProps) {
  const {
    tabs,
    currentTabIndex = I18nManager.isRTL ? tabs.length - 1 : 0,
    onTabPress,
    onTabLongPress,
  } = props;
  // Used to store the x position of each tab button
  const tabsPositionsX = useSharedValue<number[]>([0, 0, 0, 0, 0]);

  // Register each tab button's x position in the shared value
  // to be used by the indicator animation
  const handleRegisterLayout = (index: number) => (event: LayoutChangeEvent) => {
    const tabX = event.nativeEvent.layout.x;
    tabsPositionsX.modify((value) => {
      'worklet';
      value[index] = tabX;
      return value;
    });
  };

  // Animate the indicator to the current tab's x position
  const indicatorAnimatedStyles = useAnimatedStyle(() => {
    const currentX = tabsPositionsX.value[currentTabIndex];

    // On RTL languages, we need to start from -300 to 0
    const startFrom = I18nManager.isRTL ? -tabsPositionsX.value[0] : 0;

    return {
      position: 'absolute',
      transform: [{ translateX: withEaseOutQuad(startFrom + currentX) }],
    };
  });

  return (
    <AnimatedBox
      flex={1}
      flexDirection="row"
      justifyContent="space-between"
      alignItems="center"
      padding="s"
      height={TAB_BAR_HEIGHT}
      borderRadius="circle"
      maxWidth={540}
      backgroundColor="black"
      testID={'tab-bar'}>
      <Box flex={1} flexDirection="row" justifyContent="space-between" borderRadius={'circle'}>
        <Animated.View style={indicatorAnimatedStyles}>
          <Circle size={TAB_SIZE} backgroundColor="primaryInvert" />
        </Animated.View>

        {tabs.map((tab, index) => {
          const isFocused = currentTabIndex === index;

          const onPress = () => {
            if (tab.onPress) tab.onPress();
            else if (onTabPress) onTabPress(index, tab);
          };

          const onLongPress = () => {
            if (tab.onLongPress) tab.onLongPress();
            else if (onTabLongPress) onTabLongPress(index, tab);
          };

          return (
            <View key={tab.id} onLayout={handleRegisterLayout(index)}>
              <TabItem
                id={tab.id}
                label={tab.label}
                icon={tab.icon}
                focused={isFocused}
                onPress={onPress}
                onLongPress={onLongPress}
              />
            </View>
          );
        })}
      </Box>
    </AnimatedBox>
  );
}

export function routeNameToIconName(name: string) {
  return {
    index: 'home',
    shop: 'search',
    favourites: 'heart',
    bag: 'shopping-bag',
    profile: 'user',
  }[name] as FeatherIconName;
}

/**
 * Separate `TabBar` and `BottomTabBar` components has they have 2 different purposes:
 *
 * `TabBar` is a dumb component that only renders the tab bar
 *
 * `BottomTabBar`'s main concern is adpating to react-navigation's API
 * and rendering the tab bar at the bottom of the screen
 *
 * This separation allows for better testing and separation of concerns
 */
export function BottomTabBar(props: BottomTabBarProps) {
  const { state, descriptors, navigation } = props;

  const insets = useSafeAreaInsets();

  const [isBottomTabBarForcedShown] = useBottomTabBarShown();

  const currentTabIndex = state.index;

  // Determine if the tab bar should be shown
  // by checking if the current route has `tabBarShowLabel` set to `false`
  const tabBarShow = useMemo(() => {
    const currentRoute = state.routes[currentTabIndex];
    const currentDescriptor = descriptors[currentRoute.key];
    const { tabBarShowLabel } = currentDescriptor.options;

    return tabBarShowLabel || isBottomTabBarForcedShown;
  }, [state, descriptors, currentTabIndex, isBottomTabBarForcedShown]);

  const PADDING_BOTTOM = Math.max(insets.bottom, TAB_BAR_MARGIN);

  // Hide the tab bar if `tabBarShow` is set to `false`
  // by pushing it out the bottom of the screen
  const animatedContainerStyles = useAnimatedStyle(() => {
    return {
      flex: 1,
      flexDirection: 'row',
      justifyContent: 'center',
      transform: [
        {
          translateY: tabBarShow
            ? withEaseOutSin(0)
            : withEaseOutQuad(TAB_BAR_HEIGHT + PADDING_BOTTOM),
        },
      ],
    };
  });

  // Example taken from https://reactnavigation.org/docs/bottom-tab-navigator/#tabbar
  const tabs = state.routes.map((route, index) => {
    const { options } = descriptors[route.key];

    const label = (
      options.tabBarLabel !== undefined
        ? options.tabBarLabel
        : options.title !== undefined
          ? options.title
          : route.name
    ) as string;

    const isFocused = state.index === index;

    const onPress = () => {
      const event = navigation.emit({
        type: 'tabPress',
        target: route.key,
        canPreventDefault: true,
      });

      if (!isFocused && !event.defaultPrevented) {
        navigation.navigate(route.name, route.params);
      }
    };

    const onLongPress = () => {
      navigation.emit({
        type: 'tabLongPress',
        target: route.key,
      });
    };

    return {
      id: route.name,
      label,
      icon: routeNameToIconName(route.name)!,
      onPress,
      onLongPress,
    };
  });

  return (
    <View
      style={{
        position: 'absolute',
        flexDirection: 'row',
        alignItems: 'center',
        padding: TAB_BAR_MARGIN,
        paddingBottom: PADDING_BOTTOM,
        bottom: 0,
        left: 0,
        right: 0,
      }}
      testID={'bottom-tab-bar'}>
      <Animated.View style={animatedContainerStyles}>
        <TabBar tabs={tabs} currentTabIndex={currentTabIndex} />
      </Animated.View>
    </View>
  );
}

export default BottomTabBar;
