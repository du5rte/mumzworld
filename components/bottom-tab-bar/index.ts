export * from './bottom-tab-bar';
export { default } from './bottom-tab-bar';
