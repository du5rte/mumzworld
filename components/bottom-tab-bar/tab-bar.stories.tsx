import type { Meta, StoryObj } from '@storybook/react';
import { fn } from '@storybook/test';
import { useArgs } from '@storybook/preview-api';

import { TabBar } from './bottom-tab-bar';

import type { TabBarProps } from './bottom-tab-bar';
import { tabs } from './tabs-mock';

// More on how to set up stories at: https://storybook.js.org/docs/writing-stories#default-export
const meta: Meta<typeof TabBar> = {
  title: 'Formations/TabBar',
  component: TabBar,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/configure/story-layout
    layout: 'centered',
  },
  // More on argTypes: https://storybook.js.org/docs/api/argtypes
  argTypes: {
    currentTabIndex: {
      control: { type: 'range', min: 0, max: tabs.length, step: 1 },
    },
  },
  // Use `fn` to spy on the onClick arg, which will appear in the actions panel once invoked: https://storybook.js.org/docs/essentials/actions#action-args
  args: {
    tabs,
    currentTabIndex: 1,
    onTabPress: fn(),
    onTabLongPress: fn(),
  },
  // render: (args) => <WrappedRender {...args} />,
  render: function Render(args: TabBarProps) {
    const [, updateArgs] = useArgs();

    return <TabBar {...args} onTabPress={(i) => updateArgs({ currentTabIndex: i })} />;
  },
};

export default meta;

type Story = StoryObj<typeof TabBar>;

export const Default: Story = {};
