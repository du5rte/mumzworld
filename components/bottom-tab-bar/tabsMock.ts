import { Tab } from '../bottom-tab-bar';

export const tabs: Tab[] = [
  {
    id: 'index',
    icon: 'home',
    label: 'Home',
  },
  {
    id: 'shop',
    icon: 'search',
    label: 'Shop',
  },
  {
    id: 'favourites',
    icon: 'heart',
    label: 'Favourites',
  },
  {
    id: 'bag',
    icon: 'shopping-bag',
    label: 'Cart',
  },
  {
    id: 'profile',
    icon: 'user',
    label: 'Profile',
  },
];
