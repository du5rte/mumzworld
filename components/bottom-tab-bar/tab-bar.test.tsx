import React from 'react';

import { fireEvent, render, screen } from '@/utils/tests/react-native';
import { NavigationContainer } from '@react-navigation/native';
import { createBottomTabNavigator } from '@react-navigation/bottom-tabs';

import { BottomTabBar, TabBar } from './bottom-tab-bar';
import { tabs } from './tabsMock';

jest.useFakeTimers();

// Mock necessary modules
jest.mock('react-native-safe-area-context', () => ({
  useSafeAreaInsets: () => ({ bottom: 0 }),
}));
jest.mock('@react-navigation/bottom-tabs/src/types', () => ({
  // Mock other necessary parts from react-navigation
}));

describe('TabBar', () => {
  it('should contain a `tab-bar` test id', () => {
    render(<TabBar currentTabIndex={1} tabs={tabs} />);

    expect(screen.getByTestId('tab-bar')).toBeDefined();
  });

  it('should contain multiple `tab-bar-item` test ids', () => {
    render(<TabBar currentTabIndex={1} tabs={tabs} />);

    expect(screen.getAllByTestId(/tab-item/).length).toEqual(5);
  });

  it('should contain one focused `tab-item` test ids', () => {
    render(<TabBar currentTabIndex={1} tabs={tabs} />);

    expect(screen.getByTestId(/tab-item-\w+-focused/)).toBeDefined();
  });
});

describe('BottomTabBar', () => {
  const Tab = createBottomTabNavigator();

  // Helper function to setup the test environment
  const setup = () => {
    const screen = render(
      <NavigationContainer>
        <Tab.Navigator tabBar={(props) => <BottomTabBar {...props} />}>
          <Tab.Screen name="Home" component={() => null} />
          <Tab.Screen name="Settings" component={() => null} />
        </Tab.Navigator>
      </NavigationContainer>
    );
    return { screen };
  };

  it('renders correctly', () => {
    const { screen } = setup();
    expect(screen.getByTestId('bottom-tab-bar')).toBeTruthy();
  });

  it('navigates to the correct screen on tab press', () => {
    const { screen } = setup();
    const settingsTab = screen.getByTestId('tab-item-Settings');
    fireEvent.press(settingsTab);
    expect(screen.getByText('Settings')).toBeTruthy();
  });

  it('adjusts for safe area insets', () => {
    // Assuming useSafeAreaInsets returns { bottom: 34 } for this test
    jest.mock('react-native-safe-area-context', () => ({
      useSafeAreaInsets: () => ({ bottom: 34 }),
    }));
    const { screen } = setup();
    const tabBarStyle = screen.getByTestId('bottom-tab-bar').props.style;
    expect(tabBarStyle.paddingBottom).toBeGreaterThan(0);
  });

  // Add more tests as needed
});
