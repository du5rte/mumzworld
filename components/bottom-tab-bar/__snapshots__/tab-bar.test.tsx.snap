// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bottom-tab-bar should render correctly 1`] = `
<View
  collapsable={false}
  jestAnimatedStyle={
    {
      "value": {},
    }
  }
  style={
    {
      "backgroundColor": "hsl(0, 0%, 0%)",
      "borderRadius": 64,
      "flex": 1,
      "flexGrow": 1,
      "height": 64,
      "maxWidth": 400,
      "minWidth": 288,
      "padding": 8,
    }
  }
  testID="tab-bar"
>
  <View
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {
          "opacity": 1,
        },
      }
    }
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "opacity": 1,
      }
    }
  >
    <View
      collapsable={false}
      jestAnimatedStyle={
        {
          "value": {
            "position": "absolute",
            "transform": [
              {
                "translateX": 0,
              },
            ],
          },
        }
      }
      style={
        {
          "position": "absolute",
          "transform": [
            {
              "translateX": 0,
            },
          ],
        }
      }
    >
      <View
        size={48}
        style={
          [
            {
              "alignItems": "center",
              "backgroundColor": "hsl(0, 0%, 100%)",
              "borderRadius": 9999,
              "height": 48,
              "justifyContent": "center",
              "width": 48,
            },
          ]
        }
      />
    </View>
    <View
      onLayout={[Function]}
    >
      <View
        accessibilityLabel="Home"
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
        testID="tab-item-index"
      >
        <View
          size={48}
          style={
            [
              {
                "alignItems": "center",
                "borderRadius": 9999,
                "height": 48,
                "justifyContent": "center",
                "width": 48,
              },
            ]
          }
        >
          <Text />
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
    >
      <View
        accessibilityLabel="Shop"
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": true,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
        testID="tab-item-shop-focused"
      >
        <View
          size={48}
          style={
            [
              {
                "alignItems": "center",
                "borderRadius": 9999,
                "height": 48,
                "justifyContent": "center",
                "width": 48,
              },
            ]
          }
        >
          <Text />
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
    >
      <View
        accessibilityLabel="Favourites"
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
        testID="tab-item-favourites"
      >
        <View
          size={48}
          style={
            [
              {
                "alignItems": "center",
                "borderRadius": 9999,
                "height": 48,
                "justifyContent": "center",
                "width": 48,
              },
            ]
          }
        >
          <Text />
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
    >
      <View
        accessibilityLabel="Cart"
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
        testID="tab-item-bag"
      >
        <View
          size={48}
          style={
            [
              {
                "alignItems": "center",
                "borderRadius": 9999,
                "height": 48,
                "justifyContent": "center",
                "width": 48,
              },
            ]
          }
        >
          <Text />
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
    >
      <View
        accessibilityLabel="Profile"
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
        testID="tab-item-profile"
      >
        <View
          size={48}
          style={
            [
              {
                "alignItems": "center",
                "borderRadius": 9999,
                "height": 48,
                "justifyContent": "center",
                "width": 48,
              },
            ]
          }
        >
          <Text />
        </View>
      </View>
    </View>
  </View>
</View>
`;
